### PR TITLE
feat: add list view

### PR DESCRIPTION
**Pull request by sagzy**
This PR: 
- adds a list view of the locations of a map to both the dashboard and the embedded map (see screenshots below)
- adds `geocoding_pending?`, `geocoding_success?`, `geocoding_error?` instance methods to Location
- adds ``geocoding_pending`, `geocoding_success`, `geocoding_error` scopes to Location
- allows the use of `skip_geocoding` to `geocode_as_pending`
- in the dashboard, show a notification for pending locations and list them with "⏱ Geocoding in progress"
- in the dashboard, show an error for erroneous locations, list them with "⚠ Geocoding error" and offers a link to "Finalise" them manually
- adds a new seed user with pending / error locations: jorah@mapzy.io / ilovedanny
- improves icons (bootstrap-icons instead of emojis) & small things in the UX/UI

### Screenshots dashboard

Screenshot 2022-04-23 at 20 08 28
Screenshot 2022-04-23 at 20 08 13
Screenshot 2022-04-25 at 16 43 39
Screenshot 2022-04-25 at 16 44 59

### Screenshots embedded

Screenshot 2022-04-23 at 20 13 56
Screenshot 2022-04-23 at 20 14 05

### CODE REVIEW

**Review by shafy — CHANGES_REQUESTED** [review]
Looks good! Added a few small comments here and there.

**Comment by shafy** [comment]
I've also written down the following and will create issues for them, I think it makes more sense to fix these later on, and not in this PR:

- Keep drawer open when deleting locations
- Show warning if there are unfinalized locations (on the main map so that people know they need to do something)
- When adding location (or editing unfinalized location), don't show null island but rather whole map

File: app/javascript/controllers/map_controller.js [hunk 1]
```diff
@@ -127,4 +125,18 @@ export default class extends Controller {
...
+  findMarker(locationId) {
+    return this.markerAnchors.find(marker => marker.getAttribute("data-location-id") === locationId);
+  }
+
+  toggleHighlightMarker(markerElement) {
+    let svg = markerElement.firstChild;
+    let path = svg.querySelector('path');
+    let currentColor = path.getAttribute('fill');
+    let normalColor = "#E74D67";
+    let highlightColor = "#F99B46"
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/javascript/controllers/map_controller.js:138]: Fixed in 7380729

File: app/javascript/controllers/map_controller.js [hunk 2]
```diff
@@ -80,21 +77,30 @@ export default class extends Controller {
...
+      anchor.setAttribute("data-action", "side-panel#showPanel");
+      anchor.setAttribute("data-location-id", locationId);
 
       let marker = new mapboxgl.Marker({
         color: "#E74D67",
         anchor: "bottom"
       });
 
-      anchor.appendChild(marker.getElement())
+      let markerElement = marker.getElement();
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/javascript/controllers/map_controller.js:96]: Fixed in 7380729

File: app/helpers/shared_views_helper.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module SharedViewsHelper
+  # rubocop:disable Rails/HelperInstanceVariable
+  def shared_locations_path
+    if controller.controller_path.include?("dashboard")
+      send :dashboard_map_locations_path, @map
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/helpers/shared_views_helper.rb:7]: Fixed in 7380729

File: app/helpers/location_helper.rb [hunk 4]
```diff
@@ -1,7 +1,67 @@
...
+
+  def error_message
+    "Unfortunately, we were not able to geocode all of your locations.
+    Please finalize them from the list below."
+  end
+
+  def pending_message
+    "Your locations are being processed in the background.
+    This may take a couple of minutes."
+  end
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/helpers/location_helper.rb:66]: Fixed in 7380729

File: app/views/shared/_location_short.html.erb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            <%= pending_icon(size: 12) %>
+          </span>
+          <span class="text-sm text-italic pl-2">Geocoding in progress...</span>
+        </div>
+      <% elsif location.geocoding_error? %>
+        <div class="flex flex-row items-center text-mapzy-red">
+          <span class="h-3">
+            <%= error_icon(size: 12) %>
+          </span>
+          <span class="text-sm text-italic pl-2">Geocoding error</span>
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/views/shared/_location_short.html.erb:18]: Fixed in 7380729

File: app/views/shared/_side_panel.html.erb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<div data-side-panel-target="panel"
+     class="absolute w-full md:w-96 bg-white border-gray-200 border-2 md:border-r-0
+            inset-x-0 bottom-0 md:top-0 md:bottom-0 h-1/2 md:h-auto z-20 hidden">
+  <div class="w-full h-full p-8 flex flex-col overflow-y-scroll">
+    <div class="flex flex-row justify-end mb-4">
+      <a data-action="side-panel#hidePanel">
+        <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" class="bi bi-x-lg" viewBox="0 0 16 16">
+          <path fill-rule="evenodd" d="M13.854 2.146a.5.5 0 0 1 0 .708l-11 11a.5.5 0 0 1-.708-.708l11-11a.5.5 0 0 1 .708 0Z"/>
+          <path fill-rule="evenodd" d="M2.146 2.146a.5.5 0 0 0 0 .708l11 11a.5.5 0 0 0 .708-.708l-11-11a.5.5 0 0 0-.708 0Z"/>
+        </svg>
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/views/shared/_side_panel.html.erb:10]: Fixed in 7380729

File: db/seeds.rb [hunk 7]
```diff
@@ -38,3 +38,27 @@
 addresses.each_with_index do |addr, i|
   map.locations.create!(name: "Zürich Store #{i}", description: "This is a fun lil store!", address: addr)
 end
+
+# Map 4: Paris
+# Locations are pending or erroneous
+
+user = User.create!(email: "jorah@mapzy.io", password: "ilovedanny", password_confirmation: "ilovedanny")
```
[the remark this answers is not in the document]
Comment (sagzy) [reply db/seeds.rb:45]: ahah

File: spec/models/location_spec.rb [hunk 8]
```diff
@@ -82,4 +67,172 @@
...
+  describe "geocoding_success?" do
+    context "when latitude or longitude is blank" do
+      let(:location) { create(:location, latitude: nil, longitude: 1) }
+
+      it "returns false" do
+        expect(location.geocoding_success?).to be false
+      end
+    end
+
+    context "when else" do
```
[the remark this answers is not in the document]
Comment (sagzy) [reply spec/models/location_spec.rb:192]: Fixed in 7380729

File: app/models/location.rb [hunk 9]
```diff
@@ -33,11 +33,41 @@ class Location < ApplicationRecord
   validates :address, presence: true
   validates :name, presence: true
 
-  after_validation :geocode, if: :eligible_for_geocoding?
+  before_validation :geocode_as_pending, if: :skip_geocoding
```
[the remark this answers is not in the document]
Comment (sagzy) [reply app/models/location.rb:36]: Yeah good point, I was thinking about the same but could not think of a better alternative. This attribute does two things:
1. Does not call `.geocode` method
2. Mark this location as `geocoding_pending?` `true`

So I think `skip_geocoding` is a good compromise

[the remark this answers is not in the document]
Comment (sagzy) [reply app/models/location.rb:36]: Any better idea?

[the remark this answers is not in the document]
Comment (shafy) [reply app/models/location.rb:36]: Yeah not really. Let's leave it like this for now.

**Comment by sagzy** [comment]
Thanks for the review! Fixed the small comments in 7380729 and also the null island issue on adding / editing in 123ff04.

> Keep drawer open when deleting locations
This would be a good improvement indeed. Also would get rid of the ugly browser confirmation. But yes, let's do that later. I'll create an issue so that we don't forget.

> Show warning if there are unfinalized locations (on the main map so that people know they need to do something)
Good point too. I suggest: when the user has completed the import, she's redirected to the dashboard home with the side panel already open. Then not only she sees the warning, but she also knows that the side panel is where to look later on to check the status of the import. What do you think? @shafy

**Comment by shafy** [comment]
Sounds good! We already show a warning now (in my branch) but the drawer is not open. I would say let's leave the drawer opening for later and create an issue for it. I know it's not a lot, but I want to ship this now =)

**Review by shafy — APPROVED** [review]
Ok, let's get this merged!

**Pull request merged**